Extensions/DialogueTree/dialoguetools.js

```diff
@@ -0,0 +1,590 @@
+        }
+      }
+    });
+};
+
+/**
+ * Check if the Dialogue Tree is currentlyu parsing data. You can do things like disabling player movement while talking on an npc
+ */
+gdjs.dialogueTree.isRunning = function() {
+  return this.dialogueIsRunning;
+};
+
+/**
+ * Scroll the clipped text. This can be combined with a timer and user input to control how fast dialogue line text is scrolling.
+ */
+gdjs.dialogueTree.scrollClippedText = function() {
+  if (this.pauseScrolling || !this.dialogueIsRunning) return;
+
+  if (this.dialogueText) {
+    this.clipTextEnd += 1;
+  }
+};
+
+/**
+ * Check if text scrolling has completed and to prevent the user from skipping to next line before the current one has been printed fully.
+ */
+gdjs.dialogueTree.hasClippedScrollingCompleted = function() {
+  if (this.dialogueData && this.dialogueText.length) {
+    return this.clipTextEnd >= this.dialogueText.length;
+  }
+  return false;
+};
+
+/**
+ * Get the current dialogue line with a scrolling effect (recommended)
+ * Used with the scrollClippedText to achieve a classic scrolling text,as well as any <<wait>> effects to pause scrolling
+ */
+gdjs.dialogueTree.getClippedLineText = function() {
+  return this.dialogueText.length
+    ? this.dialogueText.substring(0, this.clipTextEnd)
+    : "";
+};
+
+/**
+ * Get the current complete dialogue line without using any scrolling effects.
+ * Note that using this instead getClippedLineText will skip any <<wait>> commands entirely.
+ */
+gdjs.dialogueTree.getLineText = function() {
+  this.clipTextEnd = this.dialogueText.length;
+  return this.dialogueText.length ? this.dialogueText : "";
+};
+
+/**
+ * Get the number of command parameters in a command with parameters that has been caught by a isCommandCalled condition
+ */
+gdjs.dialogueTree.commandParametersCount = function() {
+  if (this.commandParameters && this.commandParameters.length > 1) {
+    return this.commandParameters.length - 1;
+  }
+  return 0;
+};
+
+/**
+ * Get a command parameter in any command with parameters that has been caugth by a isCommandCalled condition
+ * @param {number} paramIndex The index of the parameter you want to get. If left empty, the expression will get the first parameter by default.
+ */
+gdjs.dialogueTree.getCommandParameter = function(paramIndex) {
+  if (
+    this.commandParameters &&
+    this.commandParameters.length >= paramIndex + 1
+  ) {
+    var returnedParam = this.commandParameters[paramIndex + 1];
+    return returnedParam ? returnedParam : "";
+  }
+  return "";
+};
+
+/**
+ * Catch <<commands>> and <<commands with parameters>> from the current Dialogue Line.
+ * You can trigger events in the event sheet that relate to the story you are telling throught the dialogue.
+ * @param {string} command The command you want to check for being called. Write it without the <<>>
+ */
+gdjs.dialogueTree.isCommandCalled = function(command) {
+  var { commandCalls, clipTextEnd, dialogueText } = gdjs.dialogueTree;
```

Comment:
Destructuring will probably crash the game on old Android as this is a recent JavaScript feature.

Reply:
(I fixed this in my changes)

Reply:
thank you @4ian :smiley_cat: